vectordraw/grader.py

```diff
@@ -0,0 +1,345 @@
+"""
+This module contains grading logic for Vector Drawing exercises.
+"""
+
+# pylint: disable=invalid-name
+
+import inspect
+import logging
+import math
+
+
+log = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+# Built-in check functions
+
+def _errmsg(default_message, check, vectors):
+    """
+    Return error message for `check` targeting a vector from `vectors`.
+
+    If `check` does not define a custom error message, fall back on `default_message`.
+    """
+    template = check.get('errmsg', default_message)
+    vec = vectors[check['vector']]
+    return template.format(
+        name=vec.name,
+        tail_x=vec.tail.x,
+        tail_y=vec.tail.y,
+        tip_x=vec.tip.x,
+        tip_y=vec.tip.y,
+        length=vec.length,
+        angle=vec.angle
+    )
+
+
+def _errmsg_point(default_message, check, point):
+    """
+    Return error message for `check` targeting `point`.
+
+    If `check` does not define a custom error message, fall back on `default_message`.
+    """
+    template = check.get('errmsg', default_message)
+    return template.format(name=check['point'], x=point.x, y=point.y)
+
+
+def check_presence(check, vectors):
+    """
+    Check if `vectors` contains vector targeted by `check`.
+    """
+    if check['vector'] not in vectors:
+        errmsg = check.get('errmsg', 'You need to use the {name} vector.')
+        return errmsg.format(name=check['vector'])
+
+
+def check_tail(check, vectors):
+    """
+    Check if tail of vector targeted by `check` is in correct position.
+    """
+    vec = vectors[check['vector']]
+    tolerance = check.get('tolerance', 1.0)
+    expected = check['expected']
+    dist = math.hypot(expected[0] - vec.tail.x, expected[1] - vec.tail.y)
+    if dist > tolerance:
+        return _errmsg('Vector {name} does not start at correct point.', check, vectors)
+
+
+def check_tip(check, vectors):
```

Comment:
This function has a great similarity to check_tail.  Is there a way to share code between the two?